### PR TITLE
Implement eip-6963

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/30595

Resolves 

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) Set default ethereum wallet to Brave Wallet
2) Check that window.braveEthereum and window.ethereum are created and isBraveWallet both
3) Test that https://eip6963.org/ sees Brave Wallet
4) Install Metamask
5) Check that window.ethereum is not overwritten by MetaMask and check that https://eip6963.org/ sees Brave Wallet
----
1) Set default ethereum wallet to Extensions(Brave Wallet fallback)
2) Check that window.braveEthereum and window.ethereum are created
3) Test https://eip6963.org/
4) Install Metamask
5) Check that window.ethereum is overwritten by Metamask Extension while window.braveEthereum is still isBraveWallet and https://eip6963.org/ sees Brave Wallet
----
1) Set default ethereum wallet to None
2) Check that there is no window.braveEthereum and window.ethereum
3) Check that https://eip6963.org/ doesn't see Brave Wallet

### CODE REVIEW

File: components/brave_wallet/resources/ethereum_provider.js
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/resources/ethereum_provider.js:1]: Fixed this

**Comment by darkdh** [comment]
Please be aware that we have this [issue](https://github.com/brave/brave-browser/issues/29921) to move away from `ethereum_provider.js` and this PR just add more debts to it. cc @bbondy

**Comment by bbondy** [comment]
For @darkdh 's mentioned reason, can we implement this in native instead please?

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 1]
```diff
@@ -236,6 +236,47 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, AttachIfWalletCreated) {
   EXPECT_EQ(browser()->tab_strip_model()->GetTabCount(), 1);
 }
 
+IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, EIP6369) {
```
Comment (kdenhartog) [inline renderer/test/js_ethereum_provider_browsertest.cc:239]: It would be good to add some tests for if another wallet exists to detect regressions in the event that things start interacting oddly with other extenions.

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 2]
```diff
@@ -643,4 +649,128 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
...
+}
+
+void JSEthereumProvider::AnnounceProvider() {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context =
+      render_frame()->GetWebFrame()->MainWorldScriptContext();
+
+  base::Value::Dict provider_info_value;
+  provider_info_value.Set("rDNS", "com.brave.wallet");
```
Comment (kdenhartog) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:684]:
```suggestion
  provider_info_value.Set("rdns", "com.brave.wallet");
```

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 3]
```diff
@@ -643,4 +649,128 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
...
+  provider_info_value.Set("uuid", uuid_);
+  provider_info_value.Set("name", "Brave Wallet");
```
Comment (kdenhartog) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:686]: nit: We may want to support localization of this string since it's user displayed based upon the language within the Browser

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 4]
```diff
@@ -643,4 +649,128 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
...
+                content::V8ValueConverter::Create()->ToV8Value(
+                    base::Value("info"), context),
+                std::move(info_object))
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  auto provider =
+      GetProperty(context, context->Global(), kEthereum).ToLocalChecked();
```
Comment (kdenhartog) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:712]: Does this just grab what's already injected on `window.ethereum`? In the instance where an extension has already overwritten it would this link to our provider object or theirs?

Comment (cypt4) [reply components/brave_wallet/renderer/js_ethereum_provider.cc:712]: At the current moment we rewrite window.ethereum and make it readonly with our provider

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 5]
```diff
@@ -643,4 +653,126 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
...
+          .ToLocalChecked()
+          .As<v8::Function>();
+
+  v8::Local<v8::Value> custom_event_args[] = {
+      content::V8ValueConverter::Create()->ToV8Value(
+          base::Value("eip6963:announceProvider"), context),
+      std::move(event_content)};
+
+  v8::Local<v8::Value> custom_event_value =
+      custom_event->NewInstanceWithSideEffectType(context, 2, custom_event_args)
```
Comment (darkdh) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:756]: why not using `NewInstance` directly? we don't specify type other than `kHasSideEffect`.
And argc should be 1 here, please use `sizeof(custom_event_args)`.

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 6]
```diff
@@ -643,4 +653,126 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
...
+          .ToLocalChecked();
+
+  v8::Local<v8::Function> dispatch_event =
+      GetProperty(context, context->Global(), "dispatchEvent")
+          .ToLocalChecked()
+          .As<v8::Function>();
+
+  v8::Local<v8::Value> dispatch_event_args[] = {std::move(custom_event_value)};
+
+  dispatch_event->Call(context, context->Global(), 1, dispatch_event_args)
```
Comment (darkdh) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:766]: sizeof(dispatch_event_args)

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 7]
```diff
@@ -236,7 +236,99 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, AttachIfWalletCreated) {
...
+
+  const GURL url = https_server_.GetURL("/simple.html");
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+
+  constexpr char kEvalIsBraveWallet[] = "window.ethereum.isBraveWallet";
+  EXPECT_TRUE(
+      content::EvalJs(primary_main_frame(), kEvalIsBraveWallet).ExtractBool());
+
+  std::string command = R"(
+    (async () => {
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:309]: move this to anonymous namespace to be shared with these two test suites.

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 8]
```diff
@@ -236,7 +236,99 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, AttachIfWalletCreated) {
...
+      extensions::ExtensionBuilder("MetaMask")
+          .SetID(metamask_extension_id)
+          .Build());
+  extensions::ExtensionSystem::Get(browser()->profile())
+      ->extension_service()
+      ->AddExtension(extension.get());
+
+  brave_wallet::SetDefaultEthereumWallet(
+      browser()->profile()->GetPrefs(),
+      brave_wallet::mojom::DefaultWallet::BraveWallet);
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:299]: At this point, this test doesn't mean much because we forbid `window.ethereum` overwrite by this default setting unless we can test with `BraveWalletPreferExtension`. So we can prove even with `window.ethereum` gets overwritten, eip6963 can still works. 

Probably worth considering doing read-only `window.braveEthereum` as our main usage and `window.ethereum` is just an alias to it just like what we do with `window.braveSolana`.

File: components/brave_wallet/renderer/js_ethereum_provider.cc (hunk 4 above)
Comment (darkdh) [reply components/brave_wallet/renderer/js_ethereum_provider.cc]: @kdenhartog this will always use our instance because no extensions can inject `window.ethereum` prior to us.
But that object can definitely be overwritten later by extension so I think we should consider doing `window.braveEthereum` as the main object to deal with eip6963 and `window.ethereum` which still complies to default wallet setting is just an alias to it.
see https://github.com/brave/brave-core/pull/18751#discussion_r1262881024 for more detail

File: components/brave_wallet/renderer/resource_helper.cc [hunk 9]
```diff
@@ -5,7 +5,11 @@
 
 #include "brave/components/brave_wallet/renderer/resource_helper.h"
 
+#include "base/base64.h"
+#include "base/memory/ref_counted_memory.h"
+#include "third_party/skia/include/core/SkBitmap.h"
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/renderer/resource_helper.cc:10]: SkBitmap is included in the ui/gfx/image/image.cc, so we have "calling 'AsBitmap' with incomplete return type 'SkBitmap'" without this include

File: components/brave_wallet/renderer/resource_helper.cc [hunk 10]
```diff
@@ -18,4 +22,25 @@ std::string LoadDataResource(const int id) {
...
+  if (resource_bundle.IsGzipped(id)) {
+    return resource_bundle.LoadDataResourceString(id);
+  }
+
+  auto image = resource_bundle.GetImageNamed(id);
+  if (image.IsEmpty()) {
+    return absl::nullopt;
+  }
+
+  auto encoded = base::MakeRefCounted<base::RefCountedBytes>();
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/renderer/resource_helper.cc:36]: It returns scoped_refptr. RefCountedBytes has byte array object, so it is better to keep it on the heap. We can't create unique since destructor is private, but RefCountedBytes extends RefCountedThreadSafe<RefCountedMemory>

File: components/brave_wallet/renderer/resource_helper.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/renderer/resource_helper.cc:10]: but we are not using the return value and we just pass the return value into `EncodeBGRASkBitmap`. Unless we access any members in SkBitmap, otherwise I don't think we need that header.

File: components/brave_wallet/renderer/resource_helper.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/renderer/resource_helper.cc:36]: sorry I should have be more specific. Why do we need `RefCountedBytes` here in order to use `EncodeBGRASkBitmap`?

[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/renderer/resource_helper.cc:36]: This is how it is used in other places, i think just to clear data after function exits

File: components/brave_wallet/renderer/resource_helper.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/renderer/resource_helper.cc:10]: I've just tried to switch #include "third_party/skia/include/core/SkBitmap.h" -> "ui/gfx/image/image.h"
I think we need complete type here since it returns value, not a pointer.

[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/renderer/resource_helper.cc:10]: ok, that makes sense for value return.

File: components/brave_wallet/renderer/resource_helper.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/renderer/resource_helper.cc:36]: Not all of them and I believe the place you reference from is passing the bytes between thread. ex. `//chrome/browser/ui/webui/theme_source.cc`
Most of them don't need that, see `//chrome/browser/web_applications/os_integration/icns_encoder.cc`

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 11]
```diff
@@ -164,12 +166,6 @@ void JSEthereumProvider::Install(bool allow_overwrite_window_ethereum_provider,
 
   // Check window.ethereum existence.
   v8::Local<v8::Object> global = context->Global();
-  v8::Local<v8::Value> ethereum_value =
-      global->Get(context, gin::StringToV8(isolate, kEthereum))
```
Comment (darkdh) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:163]: we need to check `kBraveEthereum` here

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 12]
```diff
@@ -75,12 +75,6 @@ void BraveWalletRenderFrameObserver::DidClearWindowObject() {
     return;
   }
 
-  auto dynamic_params = get_dynamic_params_callback_.Run();
```
Comment (darkdh) [inline renderer/brave_wallet/brave_wallet_render_frame_observer.cc:78]: don't remove this,  which is shared by Solana provider.

File: components/brave_wallet/renderer/js_ethereum_provider.cc [hunk 13]
```diff
@@ -200,13 +196,25 @@ void JSEthereumProvider::Install(bool allow_overwrite_window_ethereum_provider,
...
-    global
-        ->Set(context, gin::StringToSymbol(isolate, kEthereum), ethereum_proxy)
-        .Check();
+                           gin::StringToV8(isolate, kBraveEthereum), true);
+  }
+
+  v8::Local<v8::Value> ethereum_value =
+      global->Get(context, gin::StringToV8(isolate, kEthereum))
+          .ToLocalChecked();
+  if (install_ethereum_provider && ethereum_value->IsUndefined()) {
```
Comment (darkdh) [inline components/brave_wallet/renderer/js_ethereum_provider.cc:206]: this check is completely unnecessary,  please keep it as what it was.

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 14]
```diff
@@ -253,76 +273,33 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, EIP6369) {
...
-          .SetID(metamask_extension_id)
-          .Build());
-  extensions::ExtensionSystem::Get(browser()->profile())
-      ->extension_service()
-      ->AddExtension(extension.get());
-
   brave_wallet::SetDefaultEthereumWallet(
       browser()->profile()->GetPrefs(),
-      brave_wallet::mojom::DefaultWallet::BraveWallet);
+      brave_wallet::mojom::DefaultWallet::None);
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:329]: This is not right, with None there should be nothing injected, `window.braveEthereum` nor `window.ethereum`.
And EIP 6963 won't work in that scenario. What we need to test is install MetaMask and set the option to prefer extension and let MM overwrite `window.ethereum` to see if EIP 6963 works. 
Note that this is impossible to achieve before we add `window.braveEthereum`

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 15]
```diff
@@ -425,6 +425,25 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, NonConfigurable) {
   EXPECT_TRUE(content::EvalJs(primary_main_frame(), overwrite).ExtractBool());
 }
 
+IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest,
+                       BraveEthereum_NonConfigurable) {
+  brave_wallet::SetDefaultEthereumWallet(
+      browser()->profile()->GetPrefs(),
+      brave_wallet::mojom::DefaultWallet::None);
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:471]: same reason as above, None should be nothing injected.

File: components/safe_builtins/renderer/safe_builtins_helpers.h [hunk 16]
```diff
@@ -23,7 +24,7 @@ namespace brave {
 v8::MaybeLocal<v8::Value> LoadScriptWithSafeBuiltins(
     blink::WebLocalFrame* web_frame,
     const std::string& script,
-    const std::string& name);
+    const absl::optional<std::string>& name);
```
Comment (darkdh) [inline components/safe_builtins/renderer/safe_builtins_helpers.h:27]: let's move this to different PR, we hide all the usage together if we want to hide script names.
Also this is not required for implementing eip 6963 since you didn't add additional script to run and implement it using v8 code eventually .

**Review by darkdh — CHANGES_REQUESTED** [review]

File: components/brave_wallet/renderer/js_ethereum_provider.cc (hunk 13 above)
Comment (cypt4) [reply components/brave_wallet/renderer/js_ethereum_provider.cc:206]: If you are talking about checking whether ethereum_value->IsUndefined() - it is taken from line 170

Comment (darkdh) [reply components/brave_wallet/renderer/js_ethereum_provider.cc]: I was referring to `install_ethereum_provider` and also we should check `kBraveEthereum` instead as I mentioned in https://github.com/brave/brave-core/pull/18751#discussion_r1269683945 because `window.braveEthereum` and its alias `window.ethereum` will be installed together.

File: browser/profiles/brave_renderer_updater.cc [hunk 17]
```diff
@@ -178,7 +180,8 @@ void BraveRendererUpdater::UpdateRenderer(
 
   (*renderer_configuration)
       ->SetConfiguration(brave::mojom::DynamicParams::New(
-          brave_use_native_ethereum_wallet,
+          install_window_brave_ethereum_provider,
+          install_window_ethereum_provider,
```
[the remark this answers is not in the document]
Comment (darkdh) [reply browser/profiles/brave_renderer_updater.cc:185]: Please remove `install_window_brave_ethereum_provider` completely, as I mentioned in 
https://github.com/brave/brave-core/pull/18751#discussion_r1269849710, they are always installed together so there is no such case that we only install `window.braveEthereum` without `window.ethereum` nor install `window.ethereum` without `window.braveEthereum`

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 18]
```diff
@@ -229,14 +283,74 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, AttachIfWalletCreated) {
...
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+
+  EXPECT_TRUE(
+      content::EvalJs(primary_main_frame(), kTestEIP6963).ExtractBool());
+
+  EXPECT_EQ(browser()->tab_strip_model()->GetTabCount(), 1);
+}
+
+IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest,
+                       EIP6369_WithoutEthereumInstalled) {
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:321]: This test suite should be `EIP6369_MetaMaskAttached`

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 19]
```diff
@@ -333,6 +472,25 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest, NonConfigurable) {
   EXPECT_TRUE(content::EvalJs(primary_main_frame(), overwrite).ExtractBool());
 }
 
+IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest,
+                       BraveEthereum_NonConfigurable) {
+  brave_wallet::SetDefaultEthereumWallet(
+      browser()->profile()->GetPrefs(),
+      brave_wallet::mojom::DefaultWallet::BraveWallet);
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:479]: braveEthereum should be non configurable in both prefer extension and brave wallet setting.

**Review by darkdh — COMMENTED** [review]
We should also add a test suite in `JSEthereumProviderBrowserTest` to test under setting `None`,  there won't be both `window.braveEthereum` nor `window.ethereum` attached to normal pages.

**Review by mkarolin — APPROVED** [review]
`strings`++

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 12 above)
Comment (darkdh) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:78]: It's been moved down.

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc [hunk 20]
```diff
@@ -91,9 +85,17 @@ void BraveWalletRenderFrameObserver::DidClearWindowObject() {
...
+  auto dynamic_params = get_dynamic_params_callback_.Run();
+  if (!dynamic_params.install_window_brave_ethereum_provider &&
+      !dynamic_params.install_window_ethereum_provider &&
+      !dynamic_params.brave_use_native_solana_wallet) {
+    return;
+  }
+
+  if (web_frame->GetDocument().IsDOMFeaturePolicyEnabled(context, "ethereum")) {
     JSEthereumProvider::Install(
+        dynamic_params.install_window_brave_ethereum_provider,
```
Comment (darkdh) [inline renderer/brave_wallet/brave_wallet_render_frame_observer.cc:97]: we need early return if !install_window_brave_ethereum_provider && install_window_ethereum_provider which should not happen. We need a test case to guard it to make sure that doesn't happen by checking both `window.braveEthereum` and `window.ethereum` don't get injected.

File: components/brave_wallet/renderer/js_ethereum_provider.cc (hunk 11 above)
Comment (darkdh) [reply components/brave_wallet/renderer/js_ethereum_provider.cc:163]: `install_brave_ethereum_provider ` should also be checked here before we actually create the provider object.

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 20 above)
Comment (cypt4) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:104]: I've added a check before JSEthereumProvider::Install call

Comment (cypt4) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:104]: Added, also added install_brave_ethereum_provider check here before JSEthereumProvider::Install call

File: renderer/test/js_ethereum_provider_browsertest.cc (hunk 19 above)
Comment (cypt4) [reply renderer/test/js_ethereum_provider_browsertest.cc]: Added check for both

File: renderer/test/js_ethereum_provider_browsertest.cc (hunk 18 above)
Comment (cypt4) [reply renderer/test/js_ethereum_provider_browsertest.cc]: Done

File: renderer/brave_wallet/brave_wallet_render_frame_observer.cc (hunk 20 above)
Comment (cypt4) [reply renderer/brave_wallet/brave_wallet_render_frame_observer.cc:104]: Added additional checks to tests "DoNotAttachIfMetaMaskInstalled" and "DoNotAttachIfNoWalletCreated" to check conditions when window.ethereum is not created while window.braveEthereum is created

File: renderer/test/js_ethereum_provider_browsertest.cc [hunk 21]
```diff
@@ -259,10 +404,21 @@ IN_PROC_BROWSER_TEST_F(JSEthereumProviderBrowserTest,
...
-              std::string::npos);
+  // Check whether window.ethereum is not installed
+  {
+    std::string command = "window.ethereum.isBraveWallet";
+    EXPECT_TRUE(content::EvalJs(primary_main_frame(), command)
+                    .error.find("Cannot read properties of undefined") !=
+                std::string::npos);
+  }
+
+  // Check wheter window.braveEthereum is installed
```
Comment (darkdh) [inline renderer/test/js_ethereum_provider_browsertest.cc:415]: nit: whether

File: components/brave_wallet/renderer/js_ethereum_provider.cc (hunk 11 above)
Comment (darkdh) [reply components/brave_wallet/renderer/js_ethereum_provider.cc:163]: as I mentioned in Slack, with check before calling `Install()`, we don't need to pass `install_brave_ethereum_provider` anymore which also simplifies logic in Install()

Comment (darkdh) [reply components/brave_wallet/renderer/js_ethereum_provider.cc:163]: Also the reason that we check very early in Install for existence is because `DidClearWindowObject` will be called multiple times under some scenario and we don't want to create provider object when it is already created

**Review by darkdh — APPROVED** [review]
lgtm now

**Review by StephenHeaps — APPROVED** [review]
Opened https://github.com/brave/brave-ios/issues/7779 to support EIP-6963 on iOS. Marked as v1.58 assuming this isn't being uplifted to earlier builds? The change to `ethereum_provider.js` will break event emitting on iOS due to lack of `window.braveEthereum` at the moment, but we'll need to implement EIP-6963 in javascript on iOS as well.

**Comment by kdenhartog** [comment]
no concerns from sec perspective now

RE the test plan, we can test this against https://eip6963.org @srirambv I believe this is something you'd be testing so feel free to DM me for details if you need help with a test plan.

**Review by kdenhartog — APPROVED** [review]

**Review by StephenHeaps — APPROVED** [review]
